### ISSUE TITLE
allow caller to specify the prefix in a more flexible way when performing backup or restore

### ISSUE THREAD

**Issue by tchaikov** [scylladb/scylladb#20335]
now that we have

- backup API. see https://github.com/scylladb/scylladb/blob/e4b213f041b131f38a9d782c67152e1203bd3a7e/api/api-doc/storage_service.json#L750
- restore API. see #20305 (still pending on review)

they work fine in the sense that the related tests pass, and the restore API is able to restore the snapshot backup'ed using the backup API. but scylla-manager (SM for short) uses a different naming convention for the paths of objects stored in the remote.

- scylladb uses: `$bucket/$table_name/$snapshot_name/$sstable_component`
- SM uses: `$dc:$provider:$location_path/backup/sst/cluster/$cluster_id/dc/$dc_id/node/$node_id/keyspace/$keyspace/table/$table_name/$snapshot_version/$sstable_component`

in comparison with SM's hierarchical struture, scylladb's arrangement is more flattened.  but SM does this for a couple reasons:

- it wants to avoid name conflicts: as the same node could upload the sstable with the same sstable with the same name, but from different snapshot
- so SM can list with a given prefix to resume a backup operation
- so SM can list with a given prefix for dedup -- it keeps track of the sstable files using a manifest file with the same prefix.

also, instead of calling the "restore" API,  SM will have to download the backup'ed sstables using agent so it can perform load-and-stream for restoring vnode-based keyspaces.

so the backup and restore APIs need to allow the caller to specify both the bucket *and* the prefix, or at least the pattern of the prefix. this would allow SM to keep using the existing techniques to keep track of the progress and to restore a vnode keyspace.

this issue is created to track the design of the updated APIs allowing the caller to pass the prefix / or the pattern of prefix when performing backup and restore.

regarding the backup API, i am proposing following interface:

```json
{
          "path":"/storage_service/backup",
          "operations":[
              {
                  "method":"POST",
                  "summary":"Starts copying SSTables from a specified keyspace to a designated bucket in object storage",
                  "type":"string",
                  "nickname":"start_backup",
                  "produces":[
                      "application/json"
                  ],
                  "parameters":[
                      {
                          "name":"endpoint",
                          "description":"ID of the configured object storage endpoint to copy sstables to",
                          "required":true,
                          "allowMultiple":false,
                          "type":"string",
                          "paramType":"query"
                      },
                      {
                          "name":"prefix",
                          "description":"The prefix of the objects for the backuped sstables",
                          "required":true,
                          "allowMultiple":false,
                          "type":"string",
                          "paramType":"query"
                      },
                      {
                          "name":"keyspace",
                          "description":"Name of a keyspace to copy sstables from",
                          "required":true,
                          "allowMultiple":false,
                          "type":"string",
                          "paramType":"query"
                      },
                      {
                          "name":"snapshot",
                          "description":"Name of a snapshot to copy sstables from",
                          "required":false,
                          "allowMultiple":false,
                          "type":"string",
                          "paramType":"query"
                      }
                  ]
              }
          ]
      },
```
where, "bucket" is replaced with "prefix". the same applies to the "restore" API:

```json
{
          "path":"/storage_service/restore",
          "operations":[
              {
                  "method":"POST",
                  "summary":"Starts copying SSTables from a designated bucket in object storage to a specified keyspace",
                  "type":"string",
                  "nickname":"start_restore",
                  "produces":[
                      "application/json"
                  ],
                  "parameters":[
                      {
                          "name":"endpoint",
                          "description":"ID of the configured object storage endpoint to copy SSTables from",
                          "required":true,
                          "allowMultiple":false,
                          "type":"string",
                          "paramType":"query"
                      },
                      {
                          "name":"prefix",
                          "description":"Prefix of objects from which SSTables are restored",
                          "required":true,
                          "allowMultiple":false,
                          "type":"string",
                          "paramType":"query"
                      },
                      {
                          "name":"snapshot",
                          "description":"Name of a snapshot to copy SSTables from",
                          "required":true,
                          "allowMultiple":false,
                          "type":"string",
                          "paramType":"query"
                      },
                      {
                          "name":"keyspace",
                          "description":"Name of a keyspace to copy SSTables to",
                          "required":true,
                          "allowMultiple":false,
                          "type":"string",
                          "paramType":"query"
                      },
                      {
                          "name":"table",
                          "description":"Name of a table to copy SSTables to",
                          "required":false,
                          "allowMultiple":false,
                          "type":"string",
                          "paramType":"query"
                      }
                  ]
              }
          ]
      }
```

**Comment by tchaikov** [comment]
@xemul @kreuzerkrieg @Michal-Leszczynski does this make sense to you?

**Comment by xemul** [comment]
@tchaikov , bucket and prefix should come separately. Addressing an object in S3 have two different ways of specifying a bucket, currently we use only one -- we specify host as IP address and merge bucket with prefix in the object path. Another option (that we might want to support in the future) is to specify host as bucket.host and keep only prefix in the object path, thus we need to know bucket and prefix separately

**Comment by tchaikov** [comment]
@xemul thanks for pointing this out. i missed it. just updated the description of the issue. could you help take another look at your convenience?

**Comment by xemul** [comment]
For restore endpoint the "snapshot" parameter becomes unnecessary -- prefix is required and this is from where restore will need to copy sstables. The snapshot name is not going to be in use after this change

**Comment by tchaikov** [comment]
yes, indeed. removed from the "restore" API.

**Comment by kreuzerkrieg** [comment]
> * $dc:$provider:$location_path/backup/sst/cluster/$cluster_id/dc/$dc_id/node/$node_id/keyspace/$keyspace/table/$table_name/$snapshot_version/$sstable_component

Not sure I'm following. The `$prefix` part, does it include the bucket? if so then it maybe somewhat inconvenient since the `_client->upload_file` needs the bucket to work properly, of course it can be extracted from the `$prefix` but why just not have another param as in original PR, the `$bucket` so the `$prefix/$sstable_component` will be `$bucket/$prefix/$sstable_component`?
Just to make sure I got it right, the `$prefix` is equal to the `$dc:$provider:$location_path/backup/sst/cluster/$cluster_id/dc/$dc_id/node/$node_id/keyspace/$keyspace/table/$table_name/$snapshot_version/`?

**Comment by tchaikov** [comment]
@kreuzerkrieg yes. see also https://manager.docs.scylladb.com/stable/backup/specification.html#sst . but it should be opaque to us -- it's now up to S-M to decide where is the sstables should be stored.

**Comment by Michal-Leszczynski** [comment]
@tchaikov just to make sure, it won't be possible to:
1. backup with SM directory structure
2. restore with Scylla API (for both vnodes and tablets)

That's because with SM directory structure, we don't want to restore all of the sstables residing in `$dc:$provider:$location_path/backup/sst/cluster/$cluster_id/dc/$dc_id/node/$node_id/keyspace/$keyspace/table/$table_name/$snapshot_version/` dir as it contains sstables mixed from different backups (for deduplication purposes).
Are we ok with that?

Also, to be on the same page, the `snapshot_version` in `$dc:$provider:$location_path/backup/sst/cluster/$cluster_id/dc/$dc_id/node/$node_id/keyspace/$keyspace/table/$table_name/$snapshot_version/` is the version of the table when the snapshot was taken. This is not `snapshot_name`. Do you have the same definition for this?

**Comment by tchaikov** [comment]
> @tchaikov just to make sure, it won't be possible to:
> 
>     1. backup with SM directory structure
> 
>     2. restore with Scylla API (for both vnodes and tablets)
> 
> 
> That's because with SM directory structure, we don't want to restore all of the sstables residing in `$dc:$provider:$location_path/backup/sst/cluster/$cluster_id/dc/$dc_id/node/$node_id/keyspace/$keyspace/table/$table_name/$snapshot_version/` dir as it contains sstables mixed from different backups (for deduplication purposes). Are we ok with that?

i think i am with you on this.

> 
> Also, to be on the same page, the `snapshot_version` in `$dc:$provider:$location_path/backup/sst/cluster/$cluster_id/dc/$dc_id/node/$node_id/keyspace/$keyspace/table/$table_name/$snapshot_version/` is the version of the table when the snapshot was taken. This is not `snapshot_name`. Do you have the same definition for this?

ahh, no, not at all. so i misread the code on this. i was reading the SM's code.

anyhow, do you think the new APIs satisfy the needs of SM?

**Comment by tchaikov** [comment]
@mykaul we discussed this on [the backup and restore meeting](https://docs.google.com/document/d/1cRd_Loy3SQT3PRXn6aDsxpSteRoxxw3LNK4bL4XsP78/edit?usp=sharing). we *could* provide a command line option which forces SM to use its existing machinery to restore the backup. but i don't think this interoperability is a must have.

**Comment by tchaikov** [comment]
@Michal-Leszczynski what do you think ? ^

**Comment by Michal-Leszczynski** [comment]
> @mykaul we discussed this on [the backup and restore meeting](https://docs.google.com/document/d/1cRd_Loy3SQT3PRXn6aDsxpSteRoxxw3LNK4bL4XsP78/edit?usp=sharing). we could provide a command line option which forces SM to use its existing machinery to restore the backup so that we allow user to keep the existing SM backups and restore them after the upgrade. but i don't think scylla should support this interoperability.

@tchaikov I think that's fine, but is Scylla expected to perform sstable deduplication on its own at some point?
If so, how would it be achieved with proposed directory structure `$bucket/$table_name/$snapshot_name/$sstable_component`?

**Comment by tchaikov** [comment]
> > @mykaul we discussed this on [the backup and restore meeting](https://docs.google.com/document/d/1cRd_Loy3SQT3PRXn6aDsxpSteRoxxw3LNK4bL4XsP78/edit?usp=sharing). we could provide a command line option which forces SM to use its existing machinery to restore the backup so that we allow user to keep the existing SM backups and restore them after the upgrade. but i don't think scylla should support this interoperability.
> 
> @tchaikov I think that's fine, but is Scylla expected to perform sstable deduplication on its own at some point?

i think so.

> If so, how would it be achieved with proposed directory structure `$bucket/$table_name/$snapshot_name/$sstable_component`?

actually, we are not proposing any directory structure. the goal of this issue is to allow the caller to use whatever directory structure it wants. this is the change i wanted to have your confirmation.

back to the dedup-with-scylla requirement, since scylladb is now able to abort a backup operation. we will need to support resumption of a previous unfinished backup operation. in that case, dedup is a must-have. to support dedup, i think, if we have the luxury of using uuid-based sstable identifier, we should be able to list with given prefix to see if any object match the sstable to be uploaded, and check its size if it exists. just like how RClone does. but these checks do depend on a certain directory structure. as long as the list operation specifies a prefix which does not match enormous objects -- that is not efficient. so a table name and a version / snap name should be encoded in the key of object.

probably we can discuss further in tomorrow's meeting.

**Comment by regevran** [comment]
>  Scylla expected to perform sstable deduplication on its own at some point?

Yes, but probably the existing practices and available data are not enough.
There is an effort to find a solution. A paper/issue will probably pop up soon.

**Comment by xemul** [comment]
I propose to narrow down this issue to "backup", restore API is handled by #20461

**Comment by xemul** [comment]
@Michal-Leszczynski , does SM uploads sstables from snapshot into this "folder" on bucket:

> $dc:$provider:$location_path/backup/sst/cluster/$cluster_id/dc/$dc_id/node/$node_id/keyspace/$keyspace/table/$table_name/$snapshot_version/ 

?

If so, then backup API can just accept {endpoint, bucket, prefix, keyspace, table, snapshot} arguments. SM will then need to create several backup tasks, each with the prefix as shown above ( @tchaikov )

**Comment by tchaikov** [comment]
> I am going to ask questions regarding using this API as a drop-in replacement for rclone API used for sstable upload during the backup. They are perhaps outside of the scope of this PR, but I think that it's good to address them here.
> For completeness, here is the current full proposed Scylla backup API
> 
> ```json
>           "path":"/storage_service/backup",
>           "operations":[
>               {
>                   "method":"POST",
>                   "summary":"Starts copying SSTables from a specified keyspace to a designated bucket in object storage",
>                   "type":"string",
>                   "nickname":"start_backup",
>                   "produces":[
>                       "application/json"
>                   ],
>                   "parameters":[
>                       {
>                           "name":"endpoint",
>                           "description":"ID of the configured object storage endpoint to copy sstables to",
>                           "required":true,
>                           "allowMultiple":false,
>                           "type":"string",
>                           "paramType":"query"
>                       },
>                       {
>                           "name":"bucket",
>                           "description":"Name of the bucket to backup sstables to",
>                           "required":true,
>                           "allowMultiple":false,
>                           "type":"string",
>                           "paramType":"query"
>                       },
>                       {
>                            "name":"prefix",
>                            "description":"The prefix of the objects for the backuped sstables",
>                            "required":true,
>                            "allowMultiple":false,
>                            "type":"string",
>                            "paramType":"query"
>                       },
>                       {
>                           "name":"keyspace",
>                           "description":"Name of a keyspace to copy sstables from",
>                           "required":true,
>                           "allowMultiple":false,
>                           "type":"string",
>                           "paramType":"query"
>                       },
>                       {
>                           "name":"snapshot",
>                           "description":"Name of a snapshot to copy sstables from",
>                           "required":false,
>                           "allowMultiple":false,
>                           "type":"string",
>                           "paramType":"query"
>                       }
>                   ]
>               }
>           ]
> ```
> 
> ```json
> "name":"snapshot",
> "description":"Name of a snapshot to copy sstables from",
> "required":false,
> ```
> 
> What happens when snapshot is not specified? Are all existing snapshots uploaded? I guess this should be required in order to decrease the risk of uploading some leftovers from other backups.

currently, the "snapshot" parameter has to be specified. we marked it with `"required":false`, because we *plan* to enable scylla to create the snapshot by itself, so in future, the "snapshot" is not required. but now i think we have to specify it, otherwise we would need to inform the caller the snapshot id in a certain way. @xemul what do you think? i was revisiting our discussion at https://github.com/scylladb/scylladb/pull/19890#discussion_r1695002275 we had, back when this change was being reviewed.

> 
> There is no "table" param in the API description. I think that it is required if we want to be able to use this API as a drop-in replacement for currently used rclone API. SM stores sstables under the following path:
> 

this was one of the topics i wanted to discuss over the meeting. not sure if we had a consensus or not. but @xemul @kreuzerkrieg, please ack if we agree on this -- to add "table" parameter to the "backup" API.

> ```
> /backup/sst/cluster/<sm_generated_cluster_id>/dc/<dc>/node/<host_id>/keyspace/<keyspace>/table/<table>/<table_version_at_snapshot_time>/
> ```
> 
> This means that sstables coming from different tables should be placed in different directories. That's why SM would like to provide different "prefix" param for different tables that needs to be backed up. Without the "table" param, sstables from all snapshot-ed tables would end up in the same remote directory, and that's not compatible with the current SM backup location layout.
> 

yeah. i agree with you on this.

> Also, what does this Scylla API call return? Some Scylla task ID that can be used by SM to check progress or cancel this API call?

see https://github.com/scylladb/scylladb/issues/20461#issue-2507398058 . i am also updating this issue by replicating the explanation on the return value here.

> 
> @kreuzerkrieg does those questions make sense to you? But if the questions above are answered, proposed API would allow SM to use it as a drop-in replacement for current rclone API responsible for uploading sstables.

**Comment by xemul** [comment]
> this was one of the topics i wanted to discuss over the meeting. not sure if we had a consensus or not. but @xemul @kreuzerkrieg, please ack if we agree on this -- to add "table" parameter to the "backup" API.

ACK from me. Lets add `table` parameter and make it mandatory. With `prefix`, SM will be have a drop-in replacement for rclone without the need to feed list of files into the backup API

**Comment by Michal-Leszczynski** [comment]
> @Michal-Leszczynski , does SM uploads sstables from snapshot into this "folder" on bucket:
>> $dc:$provider:$location_path/backup/sst/cluster/$cluster_id/dc/$dc_id/node/$node_id/keyspace/$keyspace/table/$table_name/$snapshot_version/

yes, with the exception that the last directory is `$table_version_at_snapshot_time` and not `$snapshot_version`.

> If so, then backup API can just accept {endpoint, bucket, prefix, keyspace, table, snapshot} arguments. SM will then need to create several backup tasks, each with the prefix as shown above ( @tchaikov )

Yes, that's how currently SM uses the rclone API.
